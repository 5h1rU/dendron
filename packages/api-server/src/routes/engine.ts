import {
  DEngineV2,
  EngineQueryRequest,
  EngineWriteRequest,
} from "@dendronhq/common-all";
import { DendronEngine } from "@dendronhq/engine-server";
import { Request, Response, Router } from "express";
import { MemoryStore } from "../store/memoryStore";

const router = Router();

router.get("/health", async (_req: Request, res: Response) => {
  return res.json({ ok: 1 });
});

router.post("/query", async (req: Request, res: Response) => {
  const { ws, queryString, mode } = req.body as EngineQueryRequest;
  const engine = await MemoryStore.instance().get<DendronEngine>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  const { error, data } = await engine.query(queryString, mode);
  const cleanData = data.map((ent) => ent.toRawProps());
  res.json({ error, data: cleanData });
});

router.post("/write", async (req: Request, res: Response) => {
  const { ws, node, opts } = req.body as EngineWriteRequest;
  const engine = await MemoryStore.instance().get<DEngineV2>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  try {
    await engine.writeNote(node, opts);
    res.json({ error: null, data: null });
  } catch (err) {
    res.json({ error: JSON.stringify(err), data: null });
  }
});

export { router as engineRouter };
