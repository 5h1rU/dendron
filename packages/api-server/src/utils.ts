import { DEngineV2 } from "@dendronhq/common-all/src";
import { DendronEngineV2 } from "@dendronhq/engine-server";
import _ from "lodash";
import { MemoryStore } from "./store/memoryStore";

export function getWSKey(uri: string) {
  return _.trimEnd(uri, "/");
}

export async function putWS({
  ws,
  engine,
}: {
  ws: string;
  engine: DendronEngineV2;
}) {
  MemoryStore.instance().put(`ws:${getWSKey(ws)}`, engine);
}

export async function getWS({ ws }: { ws: string }) {
  const engine = await MemoryStore.instance().get<DEngineV2>(
    `ws:${getWSKey(ws)}`
  );
  if (!engine) {
    throw `No Engine: ${ws}`;
  }
  return engine;
}
