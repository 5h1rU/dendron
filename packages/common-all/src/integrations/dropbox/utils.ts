import { Note } from "../../node";
import { files } from "dropbox";
import path from "path";

export function denddronId2DxId(did: string): string {
  return `id:${did}`;
}
export function dxId2DendronId(dxId: string): string {
  return dxId.slice(3);
}

export function fileNameToTitle(name: string): string {
  const ext = path.extname(name);
  return path.basename(name, ext);
}

export function fileNameToTreePath(name: string): string[] {
  const title = fileNameToTitle(name);
  return title.split(".");
}

export function fileToNote(
  meta:
    | files.FileMetadata
    | files.FileMetadataReference
    | files.FolderMetadataReference,
  body?: string
) {
  const title = fileNameToTitle(meta.name);
  const id = dxId2DendronId(meta.id);
  const note = new Note({
    id,
    title,
    desc: "TODO",
    type: "note",
    schemaId: "-1",
    body
  });
  return note;
}
