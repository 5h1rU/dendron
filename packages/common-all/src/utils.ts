import minimatch from "minimatch";

export class DUtils {
  static minimatch = minimatch;
}
