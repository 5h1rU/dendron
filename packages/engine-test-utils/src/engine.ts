import { DEngineClientV2, DVault, WorkspaceOpts } from "@dendronhq/common-all";
import { launch } from "@dendronhq/api-server";
import { tmpDir } from "@dendronhq/common-server";
import {
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
} from "@dendronhq/common-test-utils";
import {
  createEngine as engineServerCreateEngine,
  DendronEngineClient,
  WorkspaceService,
} from "@dendronhq/engine-server";
import _ from "lodash";

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<DEngineClientV2>;

export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return engineServerCreateEngine(opts);
}

export { DEngineClientV2, DVault, WorkspaceOpts };

export async function createEngineFromServer(opts: WorkspaceOpts) {
  const port = await launch({});
  const engine: DEngineClientV2 = DendronEngineClient.create({
    port,
    vaultsv4: opts.vaults,
    ws: opts.wsRoot,
    vaults: [],
  });
  await engine.init();
  return engine;
}

async function setupWS(opts: { vaults: DVault[] }) {
  const wsRoot = tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  const vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault });
      return vault;
    })
  );
  return { wsRoot, vaults };
}

// @ts-ignore
export async function runEngineTestV5<TExtra = any>(
  func: RunEngineTestFunctionV4,
  opts: {
    preSetupHook?: SetupHookFunction;
    //postSetupHook?: PostSetupHookFunction;
    createEngine?: AsyncCreateEngineFunction;
    extra?: TExtra;
    expect: any;
    vaults?: DVault[];
    setupOnly?: boolean;
  }
): Promise<any> {
  const { preSetupHook, extra, vaults, createEngine } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    vaults: [{ fsPath: "vault1" }, { fsPath: "vault2" }],
  });
  const { wsRoot } = await setupWS({ vaults });
  await preSetupHook({ wsRoot, vaults });
  const engine = await createEngine({ wsRoot, vaults });
  const initResp = await engine.init();
  const testOpts = { wsRoot, vaults, engine, initResp, extra };
  if (opts.setupOnly) {
    return testOpts;
  }
  const results = (await func(testOpts)) || [];
  await runJestHarnessV2(results, expect);
  return { opts: testOpts, resp: undefined };
}
