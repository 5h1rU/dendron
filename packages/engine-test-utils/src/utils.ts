import { AssertUtils } from "@dendronhq/common-test-utils";
import { Git } from "@dendronhq/engine-server";
import fs from "fs-extra";
import path from "path";

export async function checkString(body: string, ...match: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      match,
    })
  ).toBeTruthy();
}

export async function checkNotInString(body: string, ...nomatch: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      nomatch,
    })
  ).toBeTruthy();
}

export class GitTestUtils {
  static async createRepoForWorkspace(wsRoot: string) {
    const git = new Git({ localUrl: wsRoot });
    await git.init();
    await git.add("dendron.yml");
    await git.commit({ msg: "init" });
  }
  static async createRepoWithReadme(root: string) {
    const git = new Git({ localUrl: root });
    await git.init();
    const readmePath = path.join(root, "README.md");
    fs.ensureFileSync(readmePath);
    await git.add(".");
    await git.commit({ msg: "init" });
  }
}
