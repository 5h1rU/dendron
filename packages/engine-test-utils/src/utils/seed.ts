import { DEngineClient, SeedCommands, SeedConfig } from "@dendronhq/common-all";
import { tmpDir, writeYAML } from "@dendronhq/common-server";
import { SeedCLICommand } from "@dendronhq/dendron-cli";
import { SeedInitMode, SeedUtils } from "@dendronhq/engine-server";
import path from "path";
import { GitTestUtils } from "./git";

export class TestSeedUtils {
  static async createSeedRegistry(opts: {
    engine: DEngineClient;
    wsRoot: string;
  }) {
    const config = await this.createSeed(opts);
    const id = SeedUtils.getSeedId(config);
    const root = tmpDir().name;
    const registryFile = path.join(root, "reg.yml");
    writeYAML(registryFile, { [id]: config });
    return { registryFile };
  }
  static async createSeed(opts: { engine: DEngineClient; wsRoot: string }) {
    const cli = new SeedCLICommand();
    const cmd = SeedCommands.INIT;
    const id = "dendron.foo";
    const seed: SeedConfig = {
      description: "",
      license: "",
      name: "foo",
      publisher: "dendron",
      repository: {
        type: "git",
        url: `file://${opts.wsRoot}`,
      },
      root: "vault",
    };
    await cli.execute({
      cmd,
      id,
      server: {} as any,
      config: seed,
      mode: SeedInitMode.CREATE_WORKSPACE,
      ...opts,
    });
    await GitTestUtils.addRepoToWorkspace(opts.wsRoot);
    return seed;
  }
}
