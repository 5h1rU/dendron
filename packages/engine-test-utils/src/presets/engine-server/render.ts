import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import { ENGINE_HOOKS } from "./utils";

const NOTES = {
  BASIC: new TestPresetEntryV4(
    async ({ engine }) => {
      const { data } = await engine.renderNote({
        id: "foo",
        format: "markdown",
      });
      return [{ actual: data, expected: "<p>foo body</p>", msg: "foo" }];
    },
    {
      preSetupHook: async (opts) => {
        return ENGINE_HOOKS.setupBasic(opts);
      },
    }
  ),
};

export const ENGINE_RENDER_PRESETS = {
  NOTES,
};
