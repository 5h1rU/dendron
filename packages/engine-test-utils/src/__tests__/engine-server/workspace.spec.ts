import { DVault, NoteProps } from "@dendronhq/common-all";
import { tmpDir } from "@dendronhq/common-server";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { DConfig, WorkspaceService } from "@dendronhq/engine-server";
import fs from "fs-extra";
import path from "path";
import { testWithEngine } from "../../engine";
import { GitTestUtils } from "../../utils";

describe("WorkspaceService", () => {
  describe("create", () => {
    test("basic", async () => {
      const wsRoot = tmpDir().name;
      const vaults = [{ fsPath: "vault1" }];
      await WorkspaceService.createWorkspace({ wsRoot, vaults });
      const gitignore = path.join(wsRoot, ".gitignore");
      expect(
        fs.readFileSync(gitignore, { encoding: "utf8" })
      ).toMatchSnapshot();
    });

    test("workspace Vault", async () => {
      const wsRoot = tmpDir().name;
      const vaults: DVault[] = [{ fsPath: "vault1", workspace: "foo" }];
      await WorkspaceService.createWorkspace({ wsRoot, vaults });
      expect(fs.existsSync(path.join(wsRoot, "foo", "vault1"))).toBeTruthy();
    });
  });

  describe("initialize", () => {
    testWithEngine("remoteVaults present", async ({ wsRoot, engine }) => {
      const root = tmpDir().name;
      await GitTestUtils.createRepoWithReadme(root);

      engine.config.vaults.push({
        fsPath: "remoteVault",
        remote: {
          type: "git",
          url: root,
        },
      });
      DConfig.writeConfig({ wsRoot, config: engine.config });

      const ws = new WorkspaceService({ wsRoot });
      const didClone = await ws.initialize({
        onSyncVaultsProgress: () => {},
        onSyncVaultsEnd: () => {},
      });
      expect(didClone).toEqual(true);
      expect(
        fs.existsSync(path.join(wsRoot, "remoteVault", "README.md"))
      ).toBeTruthy();
    });

    // testWithEngine(
    //   "remoteVaults present as workspace vault",
    //   async ({ wsRoot, engine }) => {
    //     const {wsRoot: wsRootRemote} = await runEngineTestV5(async ()=>{}, {
    //       vaults: [{
    //         fsPath: "gamma"
    //       }],
    //       expect, setupOnly: true, initGit: true})

    //     engine.config.vaults.push({
    //       fsPath: "repos/gamma",
    //       remote: {
    //         type: "git",
    //         url: wsRootRemote,
    //       },
    //     });
    //     DConfig.writeConfig({ wsRoot, config: engine.config });
    //     expect(engine.config).toMatchSnapshot();
    //     const ws = new WorkspaceService({ wsRoot });
    //     const didClone = await ws.initialize({
    //       onSyncVaultsProgress: () => { },
    //       onSyncVaultsEnd: () => { },
    //     });
    //     expect(didClone).toEqual(true);
    //     expect(
    //       fs.existsSync(path.join(wsRoot, "repos", "gamma", "dendron.yml"))
    //     ).toBeTruthy();
    //   }, {only: true }
    // );

    testWithEngine(
      "remoteVaults present, no initializeRemoteVaults",
      async ({ wsRoot, engine }) => {
        const root = tmpDir().name;
        await GitTestUtils.createRepoWithReadme(root);

        engine.config.vaults.push({
          fsPath: "remoteVault",
          remote: {
            type: "git",
            url: root,
          },
        });
        engine.config.initializeRemoteVaults = false;
        DConfig.writeConfig({ wsRoot, config: engine.config });
        const ws = new WorkspaceService({ wsRoot });
        const didClone = await ws.initialize({
          onSyncVaultsProgress: () => {},
          onSyncVaultsEnd: () => {},
        });
        expect(didClone).toBeFalsy();
        expect(
          fs.existsSync(path.join(wsRoot, "remoteVault", "README.md"))
        ).toBeFalsy();
      }
    );
  });

  testWithEngine(
    "commitAll",
    async ({ wsRoot, vaults }) => {
      await NoteTestUtilsV4.modifyNoteByPath(
        { wsRoot, vault: vaults[0], fname: "foo" },
        (note: NoteProps) => {
          note.body = note.body + "\n Foo";
          return note;
        }
      );
      await NoteTestUtilsV4.createNote({
        fname: "bar",
        vault: vaults[1],
        wsRoot,
      });
      const resp = await new WorkspaceService({
        wsRoot,
      }).commidAndAddAll();
      expect(resp.length).toEqual(2);
    },
    { initGit: true }
  );
});
