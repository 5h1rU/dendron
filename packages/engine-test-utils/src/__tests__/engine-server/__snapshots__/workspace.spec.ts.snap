// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkspaceService create basic 1`] = `
"node_modules
.dendron.*
build

"
`;
